sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreChangelog.scala

```diff
@@ -101,6 +101,13 @@ abstract class StateStoreChangelogWriter(
     fm.createAtomic(file, overwriteIfPossible = true)
   protected var compressedStream: DataOutputStream = compressStream(backingFileStream)
 
+  def writeLineage(lineage: Array[(Long, String)]): Unit = {
+    val lineageStr = lineage.map { case (version, uniqueId) =>
+      s"$version:$uniqueId"
+    }.mkString(" ")
```

Comment:
Do we do this format to make it human readable, or other purpose?
Why not just serialize an array of Scala case class?

Reply:
The actual changelog data also has our own format, e.g. `GET KEY`. This is a simple format to store / read the lineage.

Storing case class would be expensive than this right? This is fairly low level I believe custom format is fine

Reply:
I agree that custom format is fine, but it doesn't mean the format doesn't matter. This is a on-disk format, so it is permanent and we have to support forever. If we would like to come up with human readible format, why it is something separated by space, not JSON? The space separated strings are not robust. How if one day, someone decides that the unique ID will contain a space, this will break.